### DISCUSSION TITLE
arduino comp error

### DISCUSSION THREAD

**Discussion by task51erA** [mathieucarbou/MycilaWebSerial#11] [Q&A]
Hi i follword the manual to the book .. all the latest libs with dependencies .. 
two maschines 
same comp error:

c:\Users\task51er\Documents\Arduino\libraries\ESPAsyncWebServer\src\WebAuthentication.cpp: In function 'bool getMD5(uint8_t*, uint16_t, char*)':
c:\Users\task51er\Documents\Arduino\libraries\ESPAsyncWebServer\src\WebAuthentication.cpp:74:3: error: 'mbedtls_md5_starts_ret' was not declared in this scope; did you mean 'mbedtls_md5_starts'?
   74 |   mbedtls_md5_starts_ret(&_ctx);
      |   ^~~~~~~~~~~~~~~~~~~~~~
      |   mbedtls_md5_starts
c:\Users\task51er\Documents\Arduino\libraries\ESPAsyncWebServer\src\WebAuthentication.cpp:75:3: error: 'mbedtls_md5_update_ret' was not declared in this scope; did you mean 'mbedtls_md5_update'?
   75 |   mbedtls_md5_update_ret(&_ctx, data, len);
      |   ^~~~~~~~~~~~~~~~~~~~~~
      |   mbedtls_md5_update
c:\Users\task51er\Documents\Arduino\libraries\ESPAsyncWebServer\src\WebAuthentication.cpp:76:3: error: 'mbedtls_md5_finish_ret' was not declared in this scope; did you mean 'mbedtls_md5_finish'?
   76 |   mbedtls_md5_finish_ret(&_ctx, _buf);
      |   ^~~~~~~~~~~~~~~~~~~~~~
      |   mbedtls_md5_finish
c:\Users\task51er\Documents\Arduino\libraries\ESPAsyncWebServer\src\AsyncEventSource.cpp: In member function 'void AsyncEventSourceClient::_queueMessage(AsyncEventSourceMessage*)':
c:\Users\task51er\Documents\Arduino\libraries\ESPAsyncWebServer\src\AsyncEventSource.cpp:189:7: error: 'ets_printf' was not declared in this scope; did you mean 'vswprintf'?
  189 |       ets_printf("ERROR: Too many messages queued\n");
      |       ^~~~~~~~~~
      |       vswprintf
c:\Users\task51er\Documents\Arduino\libraries\ESPAsyncWebServer\src\AsyncWebSocket.cpp: In member function 'void AsyncWebSocketClient::_queueMessage(AsyncWebSocketMessage*)':
c:\Users\task51er\Documents\Arduino\libraries\ESPAsyncWebServer\src\AsyncWebSocket.cpp:549:7: error: 'ets_printf' was not declared in this scope; did you mean 'vswprintf'?
  549 |       ets_printf("ERROR: Too many messages queued\n");
      |       ^~~~~~~~~~
      |       vswprintf

exit status 1

Compilation error: exit status 1

i didnt find any clue anywhere ...

**Comment by mathieucarbou** [comment]
> Hi i follword the manual to the book .. all the latest libs with dependencies ..

There is no issue regarding the project, or the other one where you have also opened the same issue (https://github.com/ayushsharma82/WebSerial/issues/95). These 2 libraries are widely used, and have a CI system, which pass both on ESP32 and ESP8266.

You are no using the right compilation flags / boards or need to understand more how to use your IDE.

**Comment by crbyxwpzfl** [comment]
Sorry for all the edits. **tldr;** use @mathieucarbou [ESPAsyncWebServer](https://github.com/mathieucarbou/ESPAsyncWebServer) You can install this library via `arduino-cli lib install --git-url https://github.com/mathieucarbou/ESPAsyncWebServer.git`.
to enable `--git-url` you have to set the environment variable `export ARDUINO_LIBRARY_ENABLE_UNSAFE_INSTALL=true`

I encountered this error with @lacamera [ESPAsyncWebServer](https://github.com/lacamera/ESPAsyncWebServer). This is the only one available in Arduino Library manager and there for I defaulted to this lib. there is a pull request open for this though see [here](https://github.com/lacamera/ESPAsyncWebServer/pull/2).

then an other library I tried was @me-no-dev [ESPAsyncWebServer](https://github.com/me-no-dev/ESPAsyncWebServer) he addressed this already see this [issue](https://github.com/me-no-dev/ESPAsyncWebServer/issues/1414).
But there seems to be an other issue with this
<details open>
<summary>just for completeness this is the error I got</summary>

```c
/Users/***/Documents/Arduino/libraries/MycilaWebSerial/src/MycilaWebSerial.cpp: In lambda function:
/Users/***/Documents/Arduino/libraries/MycilaWebSerial/src/MycilaWebSerial.cpp:34:62: error: no matching function for call to 'AsyncWebServerRequest::beginResponse(int, const char [10], const uint8_t [2696], unsigned int)'
   34 |     AsyncWebServerResponse* response = request->beginResponse(200, "text/html", WEBSERIAL_HTML, sizeof(WEBSERIAL_HTML));
      |                                        ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/***/Documents/Arduino/libraries/MycilaWebSerial/src/MycilaWebSerial.h:14,
                 from /Users/***/Documents/Arduino/libraries/MycilaWebSerial/src/MycilaWebSerial.cpp:5:
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:250:29: note: candidate: 'AsyncWebServerResponse* AsyncWebServerRequest::beginResponse(const String&, size_t, AwsResponseFiller, AwsTemplateProcessor)'
  250 |     AsyncWebServerResponse *beginResponse(const String& contentType, size_t len, AwsResponseFiller callback, AwsTemplateProcessor templateCallback=nullptr);
      |                             ^~~~~~~~~~~~~
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:250:100: note:   no known conversion for argument 3 from 'const uint8_t [2696]' {aka 'const unsigned char [2696]'} to 'AwsResponseFiller' {aka 'std::function<unsigned int(unsigned char*, unsigned int, unsigned int)>'}
  250 |     AsyncWebServerResponse *beginResponse(const String& contentType, size_t len, AwsResponseFiller callback, AwsTemplateProcessor templateCallback=nullptr);
      |                                                                                  ~~~~~~~~~~~~~~~~~~^~~~~~~~
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:246:29: note: candidate: 'AsyncWebServerResponse* AsyncWebServerRequest::beginResponse(int, const String&, const String&)'
  246 |     AsyncWebServerResponse *beginResponse(int code, const String& contentType=String(), const String& content=String());
      |                             ^~~~~~~~~~~~~
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:246:29: note:   candidate expects 3 arguments, 4 provided
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:247:29: note: candidate: 'AsyncWebServerResponse* AsyncWebServerRequest::beginResponse(FS&, const String&, const String&, bool, AwsTemplateProcessor)'
  247 |     AsyncWebServerResponse *beginResponse(FS &fs, const String& path, const String& contentType=String(), bool download=false, AwsTemplateProcessor callback=nullptr);
      |                             ^~~~~~~~~~~~~
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:247:47: note:   no known conversion for argument 1 from 'int' to 'AsyncWebServerRequest::FS&' {aka 'fs::FS&'}
  247 |     AsyncWebServerResponse *beginResponse(FS &fs, const String& path, const String& contentType=String(), bool download=false, AwsTemplateProcessor callback=nullptr);
      |                                           ~~~~^~
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:248:29: note: candidate: 'AsyncWebServerResponse* AsyncWebServerRequest::beginResponse(File, const String&, const String&, bool, AwsTemplateProcessor)'
  248 |     AsyncWebServerResponse *beginResponse(File content, const String& path, const String& contentType=String(), bool download=false, AwsTemplateProcessor callback=nullptr);
      |                             ^~~~~~~~~~~~~
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:248:48: note:   no known conversion for argument 1 from 'int' to 'AsyncWebServerRequest::File' {aka 'fs::File'}
  248 |     AsyncWebServerResponse *beginResponse(File content, const String& path, const String& contentType=String(), bool download=false, AwsTemplateProcessor callback=nullptr);
      |                                           ~~~~~^~~~~~~
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:249:29: note: candidate: 'AsyncWebServerResponse* AsyncWebServerRequest::beginResponse(Stream&, const String&, size_t, AwsTemplateProcessor)'
  249 |     AsyncWebServerResponse *beginResponse(Stream &stream, const String& contentType, size_t len, AwsTemplateProcessor callback=nullptr);
      |                             ^~~~~~~~~~~~~
/Users/***/Documents/Arduino/libraries/ESP_Async_WebServer/src/ESPAsyncWebServer.h:249:51: note:   no known conversion for argument 1 from 'int' to 'Stream&'
  249 |     AsyncWebServerResponse *beginResponse(Stream &stream, const String& contentType, size_t len, AwsTemplateProcessor callback=nullptr);
      |                                           ~~~~~~~~^~~~~~
/Users/***/Documents/Arduino/libraries/MycilaWebSerial/src/MycilaWebSerial.cpp: In lambda function:
/Users/***/Documents/Arduino/libraries/MycilaWebSerial/src/MycilaWebSerial.cpp:41:15: error: 'class AsyncWebSocketClient' has no member named 'setCloseClientOnQueueFull'
   41 |       client->setCloseClientOnQueueFull(false);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~
```

</details>

again many thanks to @mathieucarbou for this library!

**Comment by mathieucarbou** [comment]
Please read the readme: https://github.com/mathieucarbou/MycilaWebSerial#dependencies

You need to use the ESPAsyncWebServer dependency which is maintained.
 The other ones are not.

- mathieucarbou/ESPAsyncWebServer
- mathieucarbou/AsyncTCP
